src/qmlfilesserver.cpp

```diff
@@ -0,0 +1,229 @@
+                       std::back_inserter(files),
+                       [] (auto &info) {
+
+            return QJsonObject {
+                {
+                    { u"page"_s, info.fileName() },
+                    { u"timestamp"_s, info.lastModified().toMSecsSinceEpoch() }
+                }
+            };
+        });
+
+        return files;
+    });
+
+    m_server->route("/.*", [this](const QHttpServerRequest &request) {
+        QString relativePath = request.url().path();
+
+        // drop trailing "/" or "/<number>/"
+        static const QRegularExpression trimRegex("^((/\\d+/)|/)");
+        relativePath = relativePath.replace(trimRegex, "");
+
+        bool isSafe = isRelativePathSafe(relativePath);
+
+        if (!isSafe)
+            return QHttpServerResponse(
+                "400 Directory traversal is not allowed",
+                QHttpServerResponder::StatusCode::BadRequest);
+
+        QFileInfo info(relativePath);
+
+        if (info.fileName() == "qmldir") {
+            QString path = findFirstExistingSourceDir(
+               QFileInfo(QUrl::fromPercentEncoding(relativePath.toUtf8())).path());
+
+            QString body = amendedQmldirContent(QDir(path).filePath("qmldir"));
+
+            if(body.isEmpty()) {
+                return QHttpServerResponse(
+                    "404 Not Found", QHttpServerResponder::StatusCode::NotFound);
+            }
+
+            return QHttpServerResponse(QByteArrayLiteral("text/plain"),
+                body.toUtf8(), QHttpServerResponder::StatusCode::Ok);
+        }
+
+        QString filePath = findFirstExistingFile(
+            QUrl::fromPercentEncoding(relativePath.toUtf8()));
+
+        if (filePath.isEmpty()) {
+            return QHttpServerResponse(
+                "404 Not Found", QHttpServerResponder::StatusCode::NotFound);
+        }
+
+        QFile file(filePath);
+        if (!file.open(QIODevice::ReadOnly))
+            return QHttpServerResponse(
+                "403 Forbidden", QHttpServerResponder::StatusCode::Forbidden);
+
+        static const QMimeDatabase mimeDb;
+        QFileInfo fileInfo(filePath);
+        QMimeType mime = mimeDb.mimeTypeForFile(fileInfo);
+        QByteArray body = file.readAll();
+
+        QHttpServerResponse resp(
+            mime.isValid() ? mime.name().toUtf8() :
+                QByteArrayLiteral("application/octet-stream"),
+            body, QHttpServerResponder::StatusCode::Ok);
+
+        return resp;
+    });
+
+    auto tcpserver = std::make_unique<QTcpServer>(this);
+    if (!tcpserver->listen(QHostAddress::Any, port) ||
+        !m_server->bind(tcpserver.get()))
+    {
+        qWarning() << "Failed to start server on port" << port;
+        return;
+    }
+
+    qDebug() << "Listening on port" << tcpserver->serverPort();
+    tcpserver.release();
+}
+
+QString QmlFilesServer::findFirstExistingSourceDir(const QString &relativePath) const
+{
+    for (const QString& base : m_basePaths)
```

Comment:
```suggestion
    for (const QString& base : std::as_const(m_basePaths))
```

Reply:
hmm, but we are in `const` method, `std::as_const` will do nothing there as all members are const already...